From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M0diis <modestas.kazlauskas@ktu.edu>
Date: Thu, 18 Jan 2024 17:41:26 +0200
Subject: [PATCH] Improve swing time ticking


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 18b12edbbdc2d974e2bc48c5d33848dc3bc9f976..f85c91ee6fc2b6c4ac66f183c4474987d51871f6 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2604,6 +2604,8 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     protected void updateSwingTime() {
+        if (!this.swinging && this.swingTime == 0) return; // SlimyPurpur - Improve swing time ticking
+
         int i = this.getCurrentSwingDuration();
 
         if (this.swinging) {
